Mathlib/Tactic/TermCongr.lean

```diff
@@ -0,0 +1,695 @@
+/-
+Copyright (c) 2023 Kyle Miller. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kyle Miller
+-/
+import Mathlib.Logic.Basic
+import Mathlib.Tactic.Congr!
+
+/-! # `congr(...)` congruence quotations
+
+This module defines a term elaborator for generating congruence lemmas
+from patterns written using quotation syntax.
+One can write `congr($hf $hx)` with `hf : f = f'` and `hx : x = x'` to get `f x = f' x'`.
+While in simple cases it might be possible to use `congr_arg` or `congr_fun`,
+congruence quotations are more general,
+since for example `f` could have implicit arguments, complicated dependent types,
+and subsingleton instance arguments such as `Decidable` or `Fintype`.
+
+The implementation strategy is the following:
+
+1. The pattern is elaborated twice, once with each hole replaced by the LHS
+   and again with each hole replaced by the RHS. We do not force the hole to
+   have any particular type while elaborating, but whenever we can figure out
+   a LHS or a RHS we unify.
```

Comment:
Could you clarify "but whenever we can figure out a LHS or a RHS we unify"? It isn't obvious to me yet what this means.

Reply:
That was a mysterious sentence, but I think I figured out what I meant.